### PR TITLE
Read Payment String -> Payment Selection, Bancaccount not automatically selected for Reference

### PR DESCRIPTION
**Pull request by TheBestPessimist**
https://github.com/metasfresh/metasfresh/issues/6310

### CODE REVIEW

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAO.java  2
- de.metas.payment.esr/src/test/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAOTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=5128239&bid=16854239)

File: de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAO.java [hunk 1]
```diff
@@ -64,6 +69,57 @@
 				.list();
 	}
 
+	/**
+	 * Formatting done according to https://www.gkb.ch/de/Documents/DC/Beratung-Produkte/Factsheets-Flyers/Handbuch-ESR/ESR-Handbuch-Postfinance-DE.pdf
```
Comment (metas-ts) [inline de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAO.java]:
```suggestion
	 * Explode the given {@code esrString} into a number of syntactically equivalent strings that can be matched against {@code C_BP_BankAccount.ESR_RenderedAccountNo}.
	 * Formatting done according to https://www.gkb.ch/de/Documents/DC/Beratung-Produkte/Factsheets-Flyers/Handbuch-ESR/ESR-Handbuch-Postfinance-DE.pdf
```

File: de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java [hunk 2]
```diff
@@ -95,8 +95,9 @@ protected ITranslatableString buildMessage()
 		if (targetAcct != null)
 		{
 			sb.append("Zielkonto:\n");
-			sb.append("\tEmpfaenger: ")
-					.append(targetAcct.getC_BPartner().getName());
+			 // TODO tbp: how to fix this?
```
Comment (metas-ts) [inline de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java]: how to fix what in particular; anyways, this hardcoded stuff needs fixing, sooner or later. an additional unsolved TODO is therefore not a big deal for me

File: de.metas.payment.esr/src/test/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAOTest.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+class ESRBPBankAccountDAOTest
+{
+	@Nested
+	class MatchingESRAccountNumbersTest
+	{
+		@Test
+		void esrHas9Digits()
+		{
+			final ImmutableSet<String> esrs = createMatchingESRAccountNumbers("01-1067-4");
+			esrs.forEach(s -> assertEquals(9, s.length()));
```
Comment (metas-ts) [inline de.metas.payment.esr/src/test/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAOTest.java]: there are nice AssertJ things for this; pls let me know next time you write unit tests, I can show you examples

File: de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java (hunk 2 above)
Comment (TheBestPessimist) [reply de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java]: _how to_ referred to getting the bpartner name. there's no more `C_BP_BankAccount.getC_BPartner()`, only `C_BP_BankAccount.getC_BPartner_ID()`, and i dont think loading the bpdao service and reading the name each time is a good idea.

File: de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAO.java (hunk 1 above)
Comment (TheBestPessimist) [reply de.metas.payment.esr/src/main/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAO.java]: added the explanation. now that i've read yours it makes so much sense what you wrote, but yesterday i totally couldn't figure out a good explanation.

File: de.metas.payment.esr/src/test/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAOTest.java (hunk 3 above)
Comment (TheBestPessimist) [reply de.metas.payment.esr/src/test/java/de/metas/payment/esr/api/impl/ESRBPBankAccountDAOTest.java]: i keep telling to myself i have to learn assertj, but then a new urgent task comes...

File: de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java (hunk 2 above)
Comment (metas-ts) [reply de.metas.banking/de.metas.banking.base/src/main/java/de/metas/banking/exception/BankingException.java]: Generally, i think it's not a terrible idea bc the record is cached if you load it outside of any particular transaction, which the DAO-methods usually do.
And besides: make an educated guess as to what `C_BP_BankAccount.getC_BPartner()` did under the hood 🤓

**Pull request merged**